### PR TITLE
Internal: Editor V2 - Package for communication with V1 [ED-9290]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: tests/jest/unit/assets/dev/js/editor/document/hooks/ui/settings/set-direction-mode.test.js [hunk 1]
```diff
@@ -19,8 +19,6 @@ describe( 'set-direction-mode--document/elements/settings', () => {
 			$previewContents: [ document ],
 		};
 
-		global.CustomEvent = class {};
```
Comment (StyleShit) [inline tests/jest/unit/assets/dev/js/editor/document/hooks/ui/settings/set-direction-mode.test.js:22]: IDK why it was here, it's unnecessary and caused issues with the new `dispatchEvent()` of the commands

File: packages/packages/v1-adapters/src/listener/listener.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	abortController = new AbortController();
+}
+
+function registerCommandListener(
+	command: CommandEventDescriptor['name'],
+	state: CommandEventDescriptor['state'],
+	callback: ListenerCallback
+) {
+	registerWindowEventListener( `elementor/commands/run/${ state }`, ( e ) => {
+		if ( e.type === 'command' && e.command === command ) {
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/listener/listener.ts:55]: shouldRunCallback

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: core/editor/assets/js/editor-loader-v2.js [hunk 3]
```diff
@@ -1,3 +1,9 @@
+window.__elementorEditorV1Loaded = new Promise( ( resolve, reject ) => {
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply core/editor/assets/js/editor-loader-v2.js:1]: Done

File: packages/packages/v1-adapters/src/index.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+export * from './listener';
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/packages/v1-adapters/src/index.ts:1]: Done

File: packages/packages/v1-adapters/src/listener/listener.ts (hunk 2 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listener/listener.ts:55]: Done

File: packages/packages/editor/src/init.tsx [hunk 5]
```diff
@@ -1,6 +1,9 @@
 import ReactDOM from 'react-dom';
 import Shell from './components/shell';
+import { startV1Listeners } from '@elementor/v1-adapters';
 
 export default function init( domElement: HTMLElement ): void {
+	startV1Listeners();
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply packages/packages/editor/src/init.tsx]: Done

File: core/editor/assets/js/editor-loader-v2.js [hunk 6]
```diff
@@ -1,3 +1,9 @@
+window.__elementorEditorV1Loaded = new Promise( ( resolve ) => {
```
Comment (Nevoss) [inline core/editor/assets/js/editor-loader-v2.js:1]:
```suggestion
window.__elementorEditorV1LoadingPromise = new Promise( ( resolve ) => {
```

File: packages/packages/v1-adapters/src/listeners/event-creators.ts [hunk 7]
```diff
@@ -23,6 +23,6 @@ export const windowEvent = ( event: WindowEventDescriptor['name'] ) : WindowEven
 	};
 };
 
-export const v1Init = () => {
+export const v1Ready = () => {
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/listeners/event-creators.ts:26]:
```suggestion
export const v1ReadyEvent = () => {
```

File: packages/packages/v1-adapters/src/listeners/event-creators.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { CommandEventDescriptor, WindowEventDescriptor } from './types';
+
+export const commandStart = ( command: CommandEventDescriptor['name'] ) : CommandEventDescriptor => {
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/listeners/event-creators.ts:3]:
```suggestion
export const commandStartEvent = ( command: CommandEventDescriptor['name'] ) : CommandEventDescriptor => {
```

File: packages/packages/v1-adapters/src/listeners/event-creators.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+	return {
+		type: 'command',
+		name: command,
+		state: 'before',
+	};
+};
+
+export const commandEnd = ( command: CommandEventDescriptor['name'] ) : CommandEventDescriptor => {
```
Comment (Nevoss) [inline packages/packages/v1-adapters/src/listeners/event-creators.ts:11]:
```suggestion
export const commandEndEvent = ( command: CommandEventDescriptor['name'] ) : CommandEventDescriptor => {
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: core/editor/assets/js/editor-loader-v2.js (hunk 6 above)
Comment (StyleShit) [reply core/editor/assets/js/editor-loader-v2.js:1]: Done

File: packages/packages/v1-adapters/src/listeners/event-creators.ts (hunk 7 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/event-creators.ts]: Done

File: packages/packages/v1-adapters/src/listeners/event-creators.ts (hunk 9 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/event-creators.ts]: Done

**Review by Nevoss — DISMISSED** [review]

File: packages/packages/v1-adapters/src/listeners/listeners.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
+import { normalizeEvent } from './utils';
+import { CommandEventDescriptor, EventDescriptor, ListenerCallback, WindowEventDescriptor } from './types';
+
+const callbacksByEvent = new Map<EventDescriptor['name'], ListenerCallback[]>();
+let abortController = new AbortController();
+
+export function listenTo(
+	eventDescriptors : EventDescriptor | EventDescriptor[],
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/listeners.ts:8]: remove extra space

File: packages/packages/v1-adapters/src/listeners/listeners.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	if ( ! Array.isArray( eventDescriptors ) ) {
+		eventDescriptors = [ eventDescriptors ];
+	}
+
+	eventDescriptors.forEach( ( event ) => {
+		const { type, name } = event;
+
+		switch ( type ) {
+			case 'command':
+				registerCommandListener( name, event.state, callback );
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/listeners.ts:20]: Consider creating one interface for both events (and the state might be undefined).

File: packages/packages/v1-adapters/src/listeners/listeners.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	abortController = new AbortController();
+}
+
+function registerCommandListener(
+	command: CommandEventDescriptor['name'],
+	state: CommandEventDescriptor['state'],
+	callback: ListenerCallback
+) {
+	registerWindowEventListener( `elementor/commands/run/${ state }`, ( e ) => {
+		const shouldRunCallback = e.type === 'command' && e.command === command;
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/listeners.ts:43]: why both conditions?
+
Not sure that the variable is needed..

File: packages/packages/v1-adapters/src/listeners/listeners.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+function addListener( event: EventDescriptor['name'] ) {
+	window.addEventListener(
+		event,
+		makeListener( event ),
+		{ signal: abortController.signal }
+	);
+}
+
+function makeListener( event: EventDescriptor['name'] ) : EventListener {
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/listeners.ts:71]:
```suggestion
function makeListenerCallback( event: EventDescriptor['name'] ) : EventListener {
```

File: packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+		listenTo(
+			commandEndEvent( commandToListen ),
+			callback
+		);
+
+		// Dispatch events.
+		dispatchCommandBefore( commandToListen );
+		dispatchCommandAfter( commandToListen );
+
+		dispatchCommandBefore( anotherCommand );
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts:38]: Consider adding a comment.

File: packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+		// Assert.
+		expect( callback ).toHaveBeenCalledTimes( 1 );
+		expect( callback ).toHaveBeenCalledWith( {
+			type: 'window-event',
+			event: 'test-event',
+			originalEvent: expect.any( Event ),
+		} );
+	} );
+
+	it( 'should accept an array of events', () => {
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts:73]: accept = handle/dispatch?

File: packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+		// Cleanup.
+		delete ( window as any ).__elementorEditorV1LoadingPromise;
+	} );
+
+	it( 'should throw when v1 is not loaded', async () => {
+		// Act & Assert.
+		try {
+			await dispatchOnV1Ready();
+		} catch ( e ) {
+			expect( e ).toBe( 'Elementor Editor V1 is not loaded' );
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts:169]: What will happen if there is not error? (when there is a bug).

File: packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+function dispatchWindowEvent( event: string ) {
+	window.dispatchEvent( new CustomEvent( event ) );
+}
+
+function waitForAllTicks() {
+	jest.runAllTimers();
+
+	return Promise.resolve();
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts:197]: Why?

File: packages/packages/editor/src/init.tsx [hunk 18]
```diff
@@ -1,6 +1,9 @@
 import ReactDOM from 'react-dom';
 import Shell from './components/shell';
+import { dispatchOnV1Ready } from '@elementor/v1-adapters';
 
 export default function init( domElement: HTMLElement ): void {
+	dispatchOnV1Ready();
```
Comment (rotemee) [inline packages/packages/editor/src/init.tsx:6]: Consider: init instead of dispatch?

File: packages/packages/v1-adapters/src/listeners/utils.ts [hunk 19]
```diff
@@ -0,0 +1,38 @@
+import { ListenerEvent } from './types';
+
+export function dispatchOnV1Ready() {
+	return getV1LoadingPromise().then( () => {
+		window.dispatchEvent( new CustomEvent( 'elementor/v1/initialized' ) );
+	} );
+}
+
+function getV1LoadingPromise() {
+	type ExtendedWindow = Window & {
```
Comment (rotemee) [inline packages/packages/v1-adapters/src/listeners/utils.ts:10]: Consider using ts-ignore or just change to one-line casting.

**Review by rotemee — CHANGES_REQUESTED** [review]

File: packages/packages/v1-adapters/src/listeners/listeners.ts (hunk 10 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/listeners.ts:8]: Done

File: packages/packages/v1-adapters/src/listeners/listeners.ts (hunk 11 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/listeners.ts:20]: feels kinda odd 🤷

File: packages/packages/v1-adapters/src/listeners/listeners.ts (hunk 12 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/listeners.ts:43]: both conditions are to satisfy TS, the variable is for readability (@Nevoss's idea)

File: packages/packages/v1-adapters/src/listeners/listeners.ts (hunk 13 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/listeners.ts:71]: Done

File: packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts (hunk 14 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts:38]: Done

File: packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts (hunk 15 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts:73]: Changed

File: packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts (hunk 17 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/__tests__/index.test.ts:197]: removed

File: packages/packages/editor/src/init.tsx (hunk 18 above)
Comment (StyleShit) [reply packages/packages/editor/src/init.tsx:6]: thought about it, but `dispatch` is what it actually does, WDYT?

File: packages/packages/v1-adapters/src/listeners/utils.ts (hunk 19 above)
Comment (StyleShit) [reply packages/packages/v1-adapters/src/listeners/utils.ts:10]: Done

**Review by rotemee — APPROVED** [review]

**Pull request merged**